pkg/sql/backfill.go

```diff
@@ -1233,13 +1247,26 @@ func (sc *SchemaChanger) validateForwardIndexes(
 
 			// Now compare with the row count in the table.
 			select {
-			case <-tableCountReady:
-				if idxLen != tableRowCount {
+			case <-tableCountsReady:
+				expectedCount := tableRowCount
+				// If the index is a partial index, the expected number of rows
+				// is different than the total number of rows in the table.
+				if idx.IsPartial() {
+					expectedCount = partialIndexExpectedCounts[idx.ID]
+				}
+
+				if idxLen != expectedCount {
 					// TODO(vivek): find the offending row and include it in the error.
+					// TODO(mgartner): This should be an assertion error,
```

Comment:
cc @dt

Reply:
This is the test I used. Notice that the error string does not match the one created in this block of code (`"%d entries, expected %d violates unique constraint %q"`).

```
root@127.0.0.1:53107/defaultdb> set experimental_partial_indexes=on;
SET

root@127.0.0.1:52537/defaultdb> create table t (k int primary key, a int, b int);
CREATE TABLE

root@127.0.0.1:52537/defaultdb> insert into t (k, a, b) select i, 1000000-i, 1 from
generate_series(1, 1000000) s(i);
INSERT 1000000

root@127.0.0.1:52537/defaultdb> insert into t values (1000001, 999999, 1);
INSERT 1

root@127.0.0.1:52537/defaultdb> select * from t where a = 999999;
     k    |   a    | b
----------+--------+----
        1 | 999999 | 1
  1000001 | 999999 | 1
(2 rows)

root@127.0.0.1:52537/defaultdb> create unique index i on t (a) where b > 0;
ERROR: duplicate key value (a)=(999999) violates unique constraint "i"
SQLSTATE: 23505
```

Reply:
I'm not sure that this TODO is true? Unless you set `DisallowShadowing`, if the keys are in different SSTs, we won't see the collision when we flush those SSTs, so the only way to know we violated the constraint is to go back and count. 100k entries seems like it'd easily fit in one SST so I'm not sure that that test proves this code is unreachable, but rather just that the sort-buf caught it first?

Reply:
This test has 1 million entries, but I'm not sure that makes a difference. I'm happy to remove this TODO. It'd be nice if there was a way to set the batch size to 1 in a logictest so that I could test this code path, but I'm not sure how much work that would require.

Reply:
It has been removed.